tests/e2e/framework/localregistry.go

```diff
@@ -0,0 +1,60 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package framework
+
+import (
+	"istio.io/istio/pkg/log"
+	"istio.io/istio/tests/util"
+)
+
+// LocalRegistry provides in-cluster docker registry for test
+type LocalRegistry struct {
+	namespace  string
+	istioctl   *Istioctl
+	active     bool
+	Kubeconfig string
+	file       string
+	hub        string
+	tag        string
+}
+
+// NewLocalRegistry creates a new LocalRegistry
+func NewLocalRegistry(namespace string, istioctl *Istioctl, file, kubeconfig, hub, tag string) *LocalRegistry {
+	return &LocalRegistry{
+		namespace:  namespace,
+		istioctl:   istioctl,
+		Kubeconfig: kubeconfig,
+		file:       file,
+		hub:        hub,
+		tag:        tag,
+	}
+}
+
+// Setup implements the Cleanable interface
+// Deploy the local registry to the cluster
+func (l *LocalRegistry) Setup() error {
```

Comment:
Should setup do the deployment of the local registry?

Reply:
That is the original purpose. But now the deployment process is handed over to makefile.

Reply:
So my question is: why is this being done in the makefile and not here?

Reply:
The sequence of setting up the test is: bring up local registry -> build images -> push images to registry. The last two steps are done by calling make targets, which looks not that appropriate to be called in codes. So to make the process looks more consistent, we decided to move the deployment process into make file.

Reply:
At this moment, I think a better implementation of the integration of local registry into the test framework would be detecting the existence of the registry in cluster. If yes, set active flag to be true. Do you think this makes sense? @nmittler

Reply:
> The sequence of setting up the test is: bring up local registry -> build images -> push images to registry. The last two steps are done by calling make targets, which looks not that appropriate to be called in codes. So to make the process looks more consistent, we decided to move the deployment process into make file.

The problem with putting it into the makefile is that the user can no longer run `go test`. It is definitely a goal to be as compatible with standard go tooling as possible. If we have to invoke make from within the code, it's not pretty but it will work for now.

Reply:
> At this moment, I think a better implementation of the integration of local registry into the test framework would be detecting the existence of the registry in cluster. If yes, set active flag to be true. Do you think this makes sense?

I don't think that's good enough ... we want the test framework to automatically deploy it.